File: encoder.py
import zlib
import os
from PIL import Image
import numpy as np


def file_to_image(file_path):
    try:
        # Check if the file exists
        if not os.path.isfile(file_path):
            print(f"Error: The file '{file_path}' does not exist.")
            return

        # Read file and compress
        with open(file_path, 'rb') as f:
            file_bytes = f.read()
        compressed_bytes = zlib.compress(file_bytes)

        # Get file name and extension
        file_name = os.path.basename(file_path)
        name_bytes = file_name.encode('utf-8')

        # Combine file name and compressed file bytes
        combined_bytes = len(name_bytes).to_bytes(2, 'big') + name_bytes + compressed_bytes

        # Calculate required image size
        total_bytes = len(combined_bytes)
        side_length = int(np.ceil(np.sqrt(total_bytes / 4)))
        total_pixels = side_length * side_length

        # Pad combined_bytes to fit into the image
        padded_bytes = combined_bytes + b'\x00' * (total_pixels * 4 - total_bytes)

        # Convert bytes to numpy array and reshape to RGBA
        pixel_array = np.frombuffer(padded_bytes, dtype=np.uint8).reshape((side_length, side_length, 4))

        # Create and save the image
        img = Image.fromarray(pixel_array, 'RGBA')
        output_image_path = f"{file_name}.png"
        img.save(output_image_path)

        print(f"File '{file_path}' has been successfully encoded into image '{output_image_path}'.")

        return output_image_path

    except Exception as e:
        print(f"An error occurred: {e}")

File: scripts/encoder.js
async function compressData(data) {
    return pako.deflate(data, { to: 'uint8array' });
}

document.getElementById('encodeButton').addEventListener('click', async function() {
    const fileInput = document.getElementById('fileInput');
    const file = fileInput.files[0];
    if (!file) {
        alert("Please select a file to encode.");
        return;
    }

    const reader = new FileReader();
    reader.onload = async function(event) {
        try {
            const fileName = file.name;
            const fileData = new Uint8Array(event.target.result);

            console.log("Original File Data:", fileData);

            // Compress file data
            const compressedData = await compressData(fileData);

            console.log("Compressed Data:", compressedData);

            // Encode file name and compressed data
            const nameBytes = new TextEncoder().encode(fileName);
            const nameLengthBytes = new Uint8Array(2);
            new DataView(nameLengthBytes.buffer).setUint16(0, nameBytes.length, false);

            const combinedBytes = new Uint8Array(nameLengthBytes.length + nameBytes.length + compressedData.length);
            combinedBytes.set(nameLengthBytes);
            combinedBytes.set(nameBytes, nameLengthBytes.length);
            combinedBytes.set(compressedData, nameLengthBytes.length + nameBytes.length);

            console.log("Combined Bytes:", combinedBytes);

            // Calculate required image dimensions
            const totalBytes = combinedBytes.length;
            const sideLength = Math.ceil(Math.sqrt(totalBytes / 4));
            const totalPixels = sideLength * sideLength;

            // Pad combinedBytes to fit into the image
            const paddedBytes = new Uint8Array(totalPixels * 4);
            paddedBytes.set(combinedBytes);
            paddedBytes.fill(0, combinedBytes.length);

            console.log("Padded Bytes:", paddedBytes);

            // Create a PNG image from the pixel array using UPNG
            const rgbaArray = new Uint8Array(totalPixels * 4);
            for (let i = 0; i < totalPixels; i++) {
                rgbaArray[i * 4] = paddedBytes[i * 4];       // R
                rgbaArray[i * 4 + 1] = paddedBytes[i * 4 + 1]; // G
                rgbaArray[i * 4 + 2] = paddedBytes[i * 4 + 2]; // B
                rgbaArray[i * 4 + 3] = paddedBytes[i * 4 + 3]; // A
            }

            console.log("RGBA Array:", rgbaArray);
            const pngData = UPNG.encodeLL([rgbaArray.buffer], sideLength, sideLength, 3, 1, 8);
            console.log("PNG Data:", pngData);
            const blob = new Blob([pngData], { type: 'image/png' });
            console.log("Blob:", blob);

            const downloadLink = document.createElement('a');
            downloadLink.href = URL.createObjectURL(blob);
            downloadLink.download = `${fileName}.png`;
            downloadLink.click();

            // Clear the file input field
            fileInput.value = '';
        } catch (error) {
            console.error("An error occurred:", error);
            alert("An error occurred while encoding the file.");
        }
    };

    reader.readAsArrayBuffer(file);
});